File: 20-actors/uchiwake/tests/test_off_adapter.py
#!/usr/bin/env python3
"""uchiwake 内訳 — Open Food Facts bulk-ingest adapter tests. ADR-2606081800."""
from __future__ import annotations
import json
import sys
import pathlib
import unittest

HERE = pathlib.Path(__file__).resolve().parent
ROOT = HERE.parent
sys.path.insert(0, str(ROOT / "methods"))
sys.path.insert(0, str(ROOT / "methods" / "adapters"))

import openfoodfacts as OFF  # noqa: E402
from uchiwake_edn import gtin_check_digit_ok  # noqa: E402

SAMPLE = ROOT / "data" / "ingest" / "openfoodfacts.sample.json"


class TestOffAdapter(unittest.TestCase):
    def setUp(self):
        self.records = json.loads(SAMPLE.read_text(encoding="utf-8"))
        self.datoms, self.stats = OFF.normalize_dataset(self.records)

    def test_bad_gtin_record_skipped(self):
        # 4 records in, 1 has a wrong check digit → 3 admitted, 1 skipped
        self.assertEqual(self.stats["products_ok"], 3)
        self.assertEqual(self.stats["skipped_bad_gtin"], 1)

    def test_products_keyed_on_normalized_gtin14(self):
        prods = [d for d in self.datoms if ":product/id" in d]
        self.assertEqual(len(prods), 3)
        for p in prods:
            self.assertEqual(len(p[":product/gtin"]), 14)
            self.assertTrue(gtin_check_digit_ok(p[":product/gtin"]))
            self.assertEqual(p[":product/sourcing"], ":representative")  # OFF is crowd-sourced

    def test_ingredients_become_bom_edges_with_mass(self):
        edges = [d for d in self.datoms if ":bom.edge/id" in d]
        self.assertTrue(edges)
        # Nutella sugar edge carries the 56% estimate
        sugar = [e for e in edges if e[":bom.edge/parent"] == "gtin.03017620422003"
                 and e[":bom.edge/child"] == "mat.sugar"]
        self.assertEqual(len(sugar), 1)
        self.assertAlmostEqual(sugar[0][":bom.edge/qty"], 56.0)
        self.assertEqual(sugar[0][":bom.edge/qty-unit"], "%mass")

    def test_known_ingredients_map_to_canonical_materials(self):
        mat_ids = {d[":material/id"] for d in self.datoms if ":material/id" in d}
        for canon in ("mat.sugar", "mat.cocoa", "mat.palm-oil", "mat.water", "mat.co2", "mat.milk-powder"):
            self.assertIn(canon, mat_ids)

    def test_materials_deduped_across_products(self):
        mats = [d for d in self.datoms if ":material/id" in d]
        ids = [d[":material/id"] for d in mats]
        self.assertEqual(len(ids), len(set(ids)))  # sugar appears in all 3, emitted once

    def test_output_is_valid_edn_loadable(self):
        from uchiwake_edn import _tokens, _parse  # noqa: E402
        edn = OFF._to_edn(self.datoms)
        parsed = _parse(_tokens(edn))
        self.assertIsInstance(parsed, list)
        self.assertEqual(len([r for r in parsed if isinstance(r, dict) and ":product/id" in r]), 3)


if __name__ == "__main__":
    unittest.main(verbosity=2)

File: 20-actors/uchiwake/methods/adapters/test_off_adapter.clj
#!/usr/bin/env bb
;; uchiwake 内訳 — Open Food Facts bulk-ingest adapter tests (babashka port). ADR-2606081800.
(ns uchiwake.methods.adapters.test-off-adapter
  (:require [uchiwake.methods.adapters.openfoodfacts :as off]
            [uchiwake.methods.uchiwake-edn :as ue]
            [cheshire.core :as json]
            [clojure.edn :as edn]
            [clojure.java.io :as io]
            [clojure.string :as str]
            [clojure.test :refer [deftest is are testing run-tests]]))

;; ── fixture ───────────────────────────────────────────────────────────────────
;; Resolve the sample JSON fixture the same way the Python test does:
;; tests/test_off_adapter.py sets ROOT = parent of tests/ (= actor root), then
;; SAMPLE = ROOT / "data" / "ingest" / "openfoodfacts.sample.json".
;; This file lives in methods/adapters/, so actor-root = ../../../../ from here.
(def ^:private this-file *file*)

(defn- actor-root []
  (-> this-file io/file .getAbsoluteFile
      .getParentFile   ;; adapters/
      .getParentFile   ;; methods/
      .getParentFile)) ;; uchiwake/

(def ^:private sample-path
  (delay (io/file (actor-root) "data" "ingest" "openfoodfacts.sample.json")))

(def ^:private fixture
  "Load + normalise the sample file once."
  (delay
   (let [records (json/parse-string (slurp @sample-path))]
     (off/normalize-dataset records))))

(defn- datoms [] (first @fixture))
(defn- stats  [] (second @fixture))

;; ── tests ─────────────────────────────────────────────────────────────────────

(deftest test-bad-gtin-record-skipped
  ;; 4 records in the fixture; 1 has a wrong check digit → 3 admitted, 1 skipped
  (is (= 3 (get (stats) "products_ok")))
  (is (= 1 (get (stats) "skipped_bad_gtin"))))

(deftest test-products-keyed-on-normalized-gtin14
  (let [prods (filter #(contains? % ":product/id") (datoms))]
    (is (= 3 (count prods)))
    (doseq [p prods]
      (is (= 14 (count (get p ":product/gtin"))))
      (is (ue/gtin-check-digit-ok? (get p ":product/gtin")))
      (is (= ":representative" (get p ":product/sourcing"))))))

(deftest test-ingredients-become-bom-edges-with-mass
  (let [edges  (filter #(contains? % ":bom.edge/id") (datoms))
        ;; Nutella GTIN 3017620422003 → gtin.03017620422003; sugar edge
        sugar  (filter (fn [e]
                         (and (= "gtin.03017620422003" (get e ":bom.edge/parent"))
                              (= "mat.sugar"           (get e ":bom.edge/child"))))
                       edges)]
    (is (pos? (count edges)))
    (is (= 1 (count sugar)))
    (is (< (Math/abs (- 56.0 (get (first sugar) ":bom.edge/qty" 0.0))) 0.001))
    (is (= "%mass" (get (first sugar) ":bom.edge/qty-unit")))))

(deftest test-known-ingredients-map-to-canonical-materials
  (let [mat-ids (set (map #(get % ":material/id")
                          (filter #(contains? % ":material/id") (datoms))))]
    (doseq [canon ["mat.sugar" "mat.cocoa" "mat.palm-oil"
                   "mat.water" "mat.co2" "mat.milk-powder"]]
      (is (contains? mat-ids canon) (str "expected " canon " in mat-ids")))))

(deftest test-materials-deduped-across-products
  ;; sugar appears in all 3 products but should be emitted only once
  (let [mats (filter #(contains? % ":material/id") (datoms))
        ids  (map #(get % ":material/id") mats)]
    (is (= (count ids) (count (set ids))))))

(deftest test-output-is-valid-edn-loadable
  ;; to-edn emits real EDN (keywords as bare :kw, strings quoted).
  ;; clojure.edn/read-string parses it back with keyword keys.
  (let [edn-str (off/to-edn (datoms))
        ;; Strip leading comment lines before read-string (edn reader doesn't handle ; comments)
        stripped (str/join "\n" (remove #(str/starts-with? (str/trim %) ";") (str/split-lines edn-str)))
        parsed   (edn/read-string stripped)]
    (is (vector? parsed))
    (is (= 3 (count (filter #(and (map? %) (contains? % :product/id)) parsed))))))

;; ── runner ────────────────────────────────────────────────────────────────────
(when (= *file* (System/getProperty "babashka.file"))
  (let [{:keys [fail error]} (clojure.test/run-tests 'uchiwake.methods.adapters.test-off-adapter)]
    (System/exit (if (zero? (+ fail error)) 0 1))))
